openstack/baremetalinspector/v1/introspection/results.go

```diff
@@ -0,0 +1,226 @@
+// Z from the format, if the conversion fails, it falls back to an RFC3339
+// with no Z format supported by gophercloud.
+func (r *Introspection) UnmarshalJSON(b []byte) error {
+	type tmp Introspection
+	var s struct {
+		tmp
+		StartedAt  string `json:"started_at"`
+		FinishedAt string `json:"finished_at"`
+	}
+
+	err := json.Unmarshal(b, &s)
+	if err != nil {
+		return err
+	}
+
+	*r = Introspection(s.tmp)
+
+	if s.StartedAt != "" {
+		t, err := time.Parse(time.RFC3339, s.StartedAt)
+		if err != nil {
+			t, err = time.Parse(gophercloud.RFC3339NoZ, s.StartedAt)
+			if err != nil {
+				return err
+			}
+		}
+		r.StartedAt = t
+	}
+
+	if s.FinishedAt != "" {
+		t, err := time.Parse(time.RFC3339, s.FinishedAt)
+		if err != nil {
+			t, err = time.Parse(gophercloud.RFC3339NoZ, s.FinishedAt)
+			if err != nil {
+				return err
+			}
+		}
+		r.FinishedAt = t
+	}
+
+	return nil
+}
+
+// GetIntrospectionStatusResult is the response from a Get operation.
+// Call its Extract method to interpret it as an Introspection.
+type GetIntrospectionStatusResult struct {
+	introspectionResult
+}
+
+// StartResult ...
+type StartResult struct {
+	gophercloud.ErrResult
+}
+
+// AbortResult ...
+type AbortResult struct {
+	gophercloud.ErrResult
+}
+
+// Data represents introspection data collected.
+// The format and contents of the stored data depends on the ramdisk used
+// and plugins enabled both in the ramdisk and in inspector itself.
+// This structure has been provided for basic compatibility but it
+// will need extensions
+type Data struct {
+	CPUArch       string                 `json:"cpu_arch"`
+	MACs          []string               `json:"macs"`
+	RootDisk      RootDiskType           `json:"root_disk"`
+	Interfaces    map[string]interface{} `json:"interfaces"`
+	CPUs          int                    `json:"cpus"`
+	BootInterface string                 `json:"boot_interface"`
+	MemoryMB      int                    `json:"memory_mb"`
+	IPMIAddress   string                 `json:"ipmi_address"`
+	Inventory     InventoryType          `json:"inventory"`
+	Error         string                 `json:"error"`
+	LocalGB       int                    `json:"local_gb"`
+	AllInterfaces map[string]interface{} `json:"all_interfaces"`
+}
+
+// RootDiskType ...
+type RootDiskType struct {
+	Rotational         bool        `json:"rotational"`
+	Vendor             string      `json:"vendor"`
+	Name               string      `json:"name"`
+	Hctl               interface{} `json:"hctl"`
+	WwnVendorExtension interface{} `json:"wwn_vendor_extension"`
+	WwnWithExtension   interface{} `json:"wwn_with_extension"`
+	Model              string      `json:"model"`
+	Wwn                interface{} `json:"wwn"`
+	Serial             interface{} `json:"serial"`
+	Size               int         `json:"size"`
+}
+
+// InventoryType ...
+type InventoryType struct {
+	BmcAddress   string                 `json:"bmc_address"`
+	Interfaces   []interface{}          `json:"interfaces"`
```

Comment:
I think we can create a more defined struct based off of this: 

* https://github.com/openstack/ironic-python-agent/blob/cfb62c87658e32af373278a8de655367f3b489af/ironic_python_agent/hardware.py#L635
* https://github.com/openstack/ironic-python-agent/blob/cfb62c87658e32af373278a8de655367f3b489af/ironic_python_agent/hardware.py#L291

```go
type Data struct {
  ...
  Interfaces []Interface `json:"interfaces"`
  ...
}

type Interface struct {
  Name string `json:"name"`
  MACAddress string `json:"mac_address"`
  ...
}
```

Reply:
So the class references your pointing to are part of IPA's internal programming model. Is it wise to try and mirror the programming model in a separate API's return data?

Reply:
> Is it wise to try and mirror the programming model in a separate API's return data?

What I'm looking for is a more formal structure to the output of this API suite. As mentioned in the comments earlier, what triggered this for me is the amount of `null` values in the test fixture. This means that the field is being set to `null` _somewhere_. Contrast this to a generic dictionary that didn't have the key set at all - then the `null` fields simply wouldn't exist.

Put another way, `mac_address` is being set _somewhere_. It didn't magically appear out of thin air. So if Ironic Inspector is mirroring the internal programming model of the agent in its API response, then, yes, I think it's appropriate.

If this isn't appropriate, then the Ironic Inspector API should really have a more formal model of its API response. Returning an amorphous dictionary of data is difficult to work with - how can I be sure that a field will always exist?

Reply:
One additional thought: I'm only assuming that Ironic Inspector is getting its values from `ironic-python-agent`. The values seem to match up. If this is incorrect and it's purely coincidence, then my apologies.

But as I've mentioned before, it's the existence of `null` values that has me believing that there really is a formal definition _somewhere_.

Reply:
I created an InterfaceType for the interfaces in the inventory, which is the most important part.
I think the interfaces in the main dict can stay as this for now.